File: leetcode/1221_split_a_string_in_balanced_strings.cpp
// Copyright (c) 2019 Jacopo Notarstefano

#include <cassert>
#include <string>

using namespace std;

class Solution {
 public:
    int balancedStringSplit(string s) {
        int r_count = 0, result = 0;

        for (const auto& c : s) {
            c == 'R' ? r_count++ : r_count--;
            if (r_count == 0) {
                result++;
            }
        }

        return result;
    }
};

int main() {
    auto solution = Solution();

    assert(4 == solution.balancedStringSplit("RLRRLLRLRL"));
    assert(3 == solution.balancedStringSplit("RLLLLRRRLR"));
    assert(1 == solution.balancedStringSplit("LLLLRRRR"));
}

File: leetcode/1221_split_a_string_in_balanced_strings.py
# -*- coding: utf-8 -*-


class Solution:
    def balancedStringSplit(self, s):
        r_count, result = 0, 0

        for char in s:
            r_count += 1 if char == 'R' else -1
            if r_count == 0:
                result += 1

        return result


if __name__ == '__main__':
    solution = Solution()

    assert 4 == solution.balancedStringSplit("RLRRLLRLRL")
    assert 3 == solution.balancedStringSplit("RLLLLRRRLR")
    assert 1 == solution.balancedStringSplit("LLLLRRRR")
